x/auth/account.go

```diff
@@ -115,12 +118,154 @@ func (acc *BaseAccount) SetSequence(seq int64) error {
+	scale := sdk.NewDec(blockTime.Unix() - vacc.StartTime.Unix()).Quo(sdk.NewDec(vacc.EndTime.Unix() - vacc.StartTime.Unix()))
+
+	// Add original coins unlocked by vesting schedule
+	for _, c := range vacc.OriginalVestingCoins {
+		amt := sdk.NewDecFromInt(c.Amount).Mul(scale).RoundInt()
+
+		// Must constrain with coins left in account
+		// Since some unlocked coins may have left account due to delegation
+		currentAmount := vacc.GetCoins().AmountOf(c.Denom)
+		if currentAmount.LT(amt) {
+			amt = currentAmount
+			// prevent double count of transferred coins
+			amt = amt.Sub(vacc.TransferredCoins.AmountOf(c.Denom))
+		}
+
+		// Add non-zero coins
+		if !amt.IsZero() {
+			coin := sdk.NewCoin(c.Denom, amt)
+			unlockedCoins = unlockedCoins.Plus(sdk.Coins{coin})
+		}
+	}
+
+	return unlockedCoins
+}
+
+// Implement Vesting Account. Track transfers in and out of account
+// CONTRACT: Send amounts must be negated
+func (vacc *ContinuousVestingAccount) TrackTransfers(coins sdk.Coins) {
+	vacc.TransferredCoins = vacc.TransferredCoins.Plus(coins)
+}
+
+// Implements Vesting Account. Vests all original coins after EndTime but keeps them
+// all locked until that point.
+type DelayTransferAccount struct {
+	BaseAccount
+	TransferredCoins sdk.Coins // Any received coins are sendable immediately
+
+	// All coins unlocked after EndTime
+	EndTime time.Time
+}
+
+func NewDelayTransferAccount(addr sdk.AccAddress, originalCoins sdk.Coins, endTime time.Time) DelayTransferAccount {
+	bacc := BaseAccount{
+		Address: addr,
+		Coins:   originalCoins,
+	}
+	return DelayTransferAccount{
+		BaseAccount: bacc,
+		EndTime:     endTime,
+	}
+}
+
+// Implements VestingAccount
+func (vacc DelayTransferAccount) IsVesting(blockTime time.Time) bool {
+	return blockTime.Unix() < vacc.EndTime.Unix()
+}
+
+// Implements VestingAccount. If Time < EndTime return only net transferred coins
+// Else return all coins in account (like BaseAccount)
+func (vacc DelayTransferAccount) SendableCoins(blockTime time.Time) sdk.Coins {
+	// Check if ctx.Time < EndTime
+	if blockTime.Unix() < vacc.EndTime.Unix() {
+		// Return net transferred coins
+		// If positive, then those coins are sendable
+		sendableCoins := vacc.TransferredCoins
+		for _, c := range vacc.TransferredCoins {
+			// Must constrain with coins left in account
+			// Since some unlocked coins may have left account due to delegation
+			amt := sendableCoins.AmountOf(c.Denom)
+			currentAmount := vacc.GetCoins().AmountOf(c.Denom)
+			if currentAmount.LT(amt) {
+				delta := sdk.Coin{c.Denom, amt.Sub(currentAmount)}
+				sendableCoins = sendableCoins.Minus(sdk.Coins{delta})
+			}
+		}
+		return sendableCoins
+	}
+
+	// If EndTime has passed, DelayTransferAccount behaves like BaseAccount
+	return vacc.BaseAccount.GetCoins()
+}
+
+// Implement Vesting Account. Track transfers in and out of account
+// Send amounts must be negated
+func (vacc *DelayTransferAccount) TrackTransfers(coins sdk.Coins) {
```

Comment:
In the spirit of minimizing contracts, can we have two separate methods, "TrackSend", "TrackReceive" and have these functions handle negating.